


# sam_to_bam.R
#
# Filter out mitochondrial, unmapped, and "random" chromosome reads from sam files generated by mapping with STAR.
# Convert the sam file to bam, sort the bam, and create an index file for it.
#
# These are easy things to do file by file.  This is just to conveniently do these steps for a bunch of files while
# reporting progress and time taken, and avoiding overwriting existing files.  See below for the equivalent commands at the command line.
#
# This script operates on sam files generated by mapping with STAR.
# Not bam, because Erin found that "random" chromosome reads weren't getting removed from bams along with the mito and unknown ones.
# But maybe I should try that again now that we're doing three separate commands; maybe that was the problem somehow.
#
# USAGE:
# > fn = list.files(pattern='.sam')
# > sam_to_bam(fn, fileFilteredDest='/path/to/filtered/sams', fileBamDest='/path/to/filtered/and/indexed/bams')
#
#
# TIME:
# On yasuimac/Smintheus:
#   3.5 G sam file - 4m/5.5m filtering, 5.5m/6.6m asBam()
#   5 G sam file - 5.8m filtering, 7m asBam()
#   18 G sam file - Xm/26.6m filtering, Xm/33.5m asBam()
#
#
#
########################################################
# Here are the equivalent commands at the command line:
########################################################
#
# Filter sam file (took maybe 3-4m for a 4.6G file, and cut it down to 3.6G):
# sed '/chrM/d;/random/d;/chrUn/d' samplename_mapped_Aligned.out.sam > samplename_mapped_Aligned.out.filter.sam
#
# Convert to bam:
# samtools view -b samplename_mapped_Aligned.out.filter.sam > samplename_mapped_Aligned.out.filter.bam
#
# Sort the bam file (it will create a bunch of temporary files while it's working, and then merge them):
# samtools sort samplename_mapped_Aligned.out.filter.bam samplename_mapped_Aligned.out.filter.sorted -@ 4
#
# Index the bam file:
# samtools index samplename_mapped_Aligned.out.filter.sorted.bam


sam_to_bam = function(filenames, fileFilteredSuffix='.filtered.sam', fileBamSuffix='.filtered', fileFilteredDest='./', fileBamDest='./', filterVerbose=FALSE) {
    
    # Won't be necessary when this is in package
    library('Rsamtools')
    # source('/Users/nelsonlab/Documents/Toolboxes/code/file_checks.R')
    # source('/Users/nelsonlab/Documents/Toolboxes/code/sam_filter.R')
    source('/Volumes/CodingClub1/RNAseq/code/file_checks.R')
    source('/Volumes/CodingClub1/RNAseq/code/sam_filter.R')
    
    # Check that destination directories exist
    if ( !dir.exists(fileFilteredDest) ) {stop(paste('Directory', fileFilteredDest, 'does not exist.'))}
    if ( !dir.exists(fileBamDest) ) {stop(paste('Directory', fileBamDest, 'does not exist.'))}
    # Make sure they have a / at the end
    if ( ! substr(fileFilteredDest, nchar(fileFilteredDest), nchar(fileFilteredDest))=='/') { fileFilteredDest = paste(fileFilteredDest, '/', sep='') }
    if ( ! substr(fileBamDest, nchar(fileBamDest), nchar(fileBamDest))=='/') { fileBamDest = paste(fileBamDest, '/', sep='') }
    
    for (f in filenames) {
        
        writeLines(paste('\nProcessing file:', f))
        flush.console()
        
        # Make sure file exists and has extension .sam
        if ( ! file_checks(f, extension = '.sam') ) { next }
        flush.console()
        
        ###########################################################################
        # Remove mitochondrial, unmapped, and random chromosomes with sam_filter().
        ###########################################################################
        writeLines('Filtering...', sep='')
        flush.console()
        # Define name of filtered file and check that it doesn't already exist
        fileFiltered = sub('.sam', fileFilteredSuffix, f)
        fileFilteredFull = paste(fileFilteredDest, fileFiltered, sep='')
        if ( file_checks(fileFilteredFull, shouldExist=FALSE) ){
            # Announce output file name
            writeLines(paste('file will be saved as', fileFilteredFull), sep='')
            flush.console()
            # Call and time sam_filter()
            tStartFilter = proc.time()[3]
            sam_filter(f, fileFilteredFull, verbose = filterVerbose)
            tElapsedFilter = proc.time()[3] - tStartFilter
            # Announce you're done
            writeLines(paste('...done (', round(tElapsedFilter/60, digits=2), 'm).', sep=''))
            flush.console()
        }
        
        ###############################################################################
        # Convert to sam to bam, sort, and generate index file.  All donw with asBam().
        ###############################################################################
        writeLines('Converting to bam, sorting, and generating index file...', sep='')
        flush.console()
        # Define name of bam file and check that it doesn't already exist
        fileBamFull = paste(fileBamDest, sub(fileFilteredSuffix, fileBamSuffix, fileFiltered), sep='')
        if ( file_checks(fileBamFull, shouldExist=FALSE) ) {
            # Announce output file names
            flush.console(); writeLines(paste('file will be saved as', fileBamFull, '.bam', sep=''), sep='')
            flush.console()
            # Call and time asBam()
            tStartBam = proc.time()[3]
            # asBam() automatically adds .bam to the end of the file, so get rid of that if it's there
            asBam(fileFilteredFull, destination = fileBamFull)
            tElapsedBam = proc.time()[3] - tStartBam
            # Announce you're done
            writeLines(paste('...done (', round(tElapsedBam/60, digits=2), 'm).', sep=''))
            flush.console()
        }
    
        writeLines('Done with file. Don\'t be concerned by any warning messages concerning merging files; asBam() does that and it\'s fine.')
    
    }

}






