#' Map reads to genome
#'
#' Map reads to genome using STAR. Most parameter descriptions are from STAR manual.
#' @param readFilesIn Character - name(s) (with path) of the files containing the sequences to be mapped (e.g. RNA-seq FASTQ files)
#' @param genomeDir String - Path to directory where genome indices were generated by STAR
#' @param starPath String - Path to directory with STAR executable
#' @param outDest String - Directory where output files should be saved
#' @param outSuffix String - will be appended to original filename
#' @param runThreadN Numeric - How many cores to use
#' @param settings String - overrides values for all the following parameters by using ENCODE settings for long or short RNA
#' @param outSAMtype String - format for output alignment files (SAM, BAM, BAM SortedByCoordinate, BAM Unsorted SortedByCoordinate)
#' @param outFilterType String - reduces the number of "spurious" junctions
#' @param outFilterMultimapNmax Numeric - max number of multiple alignments allowed for a read: if exceeded, the read is considered unmapped
#' @param alignSJoverhangMin Numeric - minimum overhang for unannoted junctions
#' @param alignSJDBoverhangMin Numeric - minimum overhang for annotated junctions
#' @param outFilterMismatchNmax Numeric - maximum number of mismatches per pair, large number switches off this filter
#' @param outFilterMismatchNoverReadLmax Numeric - max number of mismatches per pair relative to read length: for 2x100b, max number of mismatches is 0.04*200=8 for the paired read
#' @param alignIntronMin Numeric - minimum intron length
#' @param alignIntronMax Numeric - maximum intron length
#' @param alignMatesGapMax Numeric - maximum genomic distance between mates
#' @param quantMode String - Get either read counts or transcript coordinates ("GeneCounts" or "TranscriptomeSAM")
#' @details Take a list of fastq files containing reads, and get alignments with STAR.  Default parameter values are defaults for STAR 2.5.3.
#' See manual for details on those.  Override by defining "settings" as "ENCODE_long" or "ENCODE_short" to use ENCODE settings for long or short RNA.
#' If an input file doesn't exist, you'll get an error.  If an output file (at least a Log.out file) does exist, it'll just skip the corresponding input file.
#' TIME:
#' Roughly 10m per 5G data file.
#' 5h 30m for all 28 RORb data files.
#' 4h 45m for the 15 EMX-Cre data files.
#' 3hr 30m for the 14 PV data files.
#' ~30m for the Nuc-seq data (12 files?), minus one file.
#'  Example at the command line using ENCODE settings for long RNA
#' (if you want to play with the parameters while looking at just one file, this might be easiest):
#' /opt/STAR/bin/MacOSX_x86_64/STAR --genomeDir $genomeDir --readFilesIn $pathTrimmed$sn$trimmedSuffix --outFilterType BySJout --outFilterMultimapNmax 20 --alignSJoverhangMin 8 --alignSJDBoverhangMin 1 --outFilterMismatchNmax 999 --outFilterMismatchNoverLmax 0.04 --alignIntronMin 20 --alignIntronMax 1000000 --alignMatesGapMax 1000000 --outSAMtype 'BAM SortedByCoordinate' --outFileNamePrefix $pathMapped$sn$mappedSuffix#'
#' @examples
#' gdir = '/Volumes/CodingCLub1/STAR_stuff/indexes/refGene_gtf_maxLen75/'
#' fIn=list.files(paste(projectPath, "trimmed/", sep=""), pattern='fastq')
#' STAR_run(paste(projectPath, "trimmed/", fIn, sep=""), outDest=paste(projectPath, "mapped/", sep=""), genomeDir = gdir, runThreadN = 8, settings = "ENCODE_long", outSAMtype = 'BAM SortedByCoordinate')
#' @author Emma Myers
#' @export


STAR_run = function(readFilesIn, genomeDir, starPath="/opt/STAR/bin/MacOSX_x86_64/", outDest="./", outSuffix="", runThreadN=1, settings=NULL,
    outSAMtype="SAM", outFilterType="Normal", outFilterMultimapNmax=10, alignSJoverhangMin=5, alignSJDBoverhangMin=3,
    outFilterMismatchNmax=10, outFilterMismatchNoverReadLmax=1, alignIntronMin=21, alignIntronMax=0, alignMatesGapMax=0,
    quantMode=NULL) {


    # Check arguments
    if ( !all(file.exists(readFilesIn)) ) {
        writeLines("Missing input files:")
        writeLines(readFilesIn[which(!file.exists(readFilesIn))])
        stop("Missing input file(s).  See above.")
    }
    if (! genomeDir == "") { genomeDir = dir_check(genomeDir) }
    if (! starPath == "") { starPath = dir_check(starPath) }
    outDest = dir_check(outDest)

    # Check if settings requested
    if ( !is.null(settings) ) {
        if (settings == "ENCODE_long") {
            outFilterType='BySJout'
            outFilterMultimapNmax=20
            alignSJoverhangMin=8  # for paired reads?
            alignSJDBoverhangMin=1  # for paired reads?
            outFilterMismatchNmax=999
            outFilterMismatchNoverLmax=0.04
            alignIntronMin=20   # for paired reads?
            alignIntronMax=1000000  # for paired reads?
            alignMatesGapMax=1000000 # for paired reads?
        } else if (settings == "ENCODE_short") {
            outFilterMismatchNoverLmax=0.03
            outFilterMultimapNmax=20  #same as long RNA
            alignIntronMax=1
            outFilterMatchNmin=16
            outFilterScoreMinOverLread=0
            outFilterMatchNminOverLread=0
        } else { stop("Invalid value for settings parameter") }
    }


    for (f in readFilesIn) {

        writeLines(paste("\nProcessing file:", f))

        fOut = paste(outDest, gsub(paste(".", tools::file_ext(f), sep=""), "", basename(f)), outSuffix, "_", sep="")
        writeLines(fOut)
        # Check if there's already a log file for this file
        if ( file_checks(paste(fOut, "Log.out", sep=""), shouldExist=FALSE, verbose=TRUE) ) {

            # Define arguments to the STAR command
            arguments = c("--readFilesIn", f, "--genomeDir", genomeDir, "--outFileNamePrefix", fOut,
                          "--runThreadN", runThreadN, "--outSAMtype", outSAMtype,
                          "--outFilterType", outFilterType, "--outFilterMultimapNmax", outFilterMultimapNmax,
                          "--alignSJoverhangMin", alignSJoverhangMin, "--alignSJDBoverhangMin", alignSJDBoverhangMin,
                          "--outFilterMismatchNmax", outFilterMismatchNmax, "--outFilterMismatchNoverReadLmax", outFilterMismatchNoverReadLmax,
                          "--alignIntronMin", alignIntronMin, "--alignIntronMax", alignIntronMax, "--alignMatesGapMax", alignMatesGapMax)
            if ( !is.null(quantMode) ) { arguments = c(arguments, "--quantMode", quantMode) }

            # Do the mapping (STAR displays time taken, so don't bother with that)
            system2( paste(starPath, "STAR", sep=""), args = arguments )

        } else {  writeLines("Skipping file.") }

    }

}

